runtime/interpreter/interpreter_expression.go

```diff
@@ -304,15 +304,25 @@ func (interpreter *Interpreter) memberExpressionGetterSetter(
 // e.g.1: Given type T, this method returns &T.
 // e.g.2: Given T?, this returns (&T)?
 func (interpreter *Interpreter) getReferenceValue(value Value, resultType sema.Type, locationRange LocationRange) Value {
+
+	// `resultType` is always an [optional] reference.
+	// This is guaranteed by the checker.
+	referenceType, ok := sema.UnwrapOptionalType(resultType).(*sema.ReferenceType)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
 	switch value := value.(type) {
 	case NilValue, ReferenceValue:
 		// Reference to a nil, should return a nil.
 		// If the value is already a reference then return the same reference.
 		// However, we need to make sure that this reference is actually a subtype of the resultType,
 		// since the checker may not be aware that we are "short-circuiting" in this case
+		// Additionally, it is only safe to "compress" reference types like this when the desired
+		// result reference type is unauthorized
 
 		staticType := value.StaticType(interpreter)
-		if !interpreter.IsSubTypeOfSemaType(staticType, resultType) {
+		if referenceType.Authorization != sema.UnauthorizedAccess || !interpreter.IsSubTypeOfSemaType(staticType, resultType) {
```

Comment:
Should we not allow if there are entitlements, but there are the same/equivalent?

Reply:
No, this is exactly the scenario that resulted in the issue in the first place. Different entitlements are already disallowed by the subtyping check, but in the example that motivated this, the entitlement mapping is pre-computed to ask for the entitlements in the output of the map.